### ISSUE TITLE
hangs when saving to pickle

### ISSUE THREAD

**Issue by hasakura511** [trevorstephens/gplearn#2]
Hi,

Thanks for creating this wonderful tool. I was wondering if saving models using pickle is supported? 

When i issue the following command:

gp = SymbolicRegressor()
trained_models = {}
trained_models['gp'] = pickle.dumps(gp)

my IDE hangs.  This usually works when I am using other sklearn models. I am using the following:

Python 2.7.10 |Anaconda 2.3.0 (64-bit)| (default, Oct 19 2015, 18:04:42) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.0 -- An enhanced Interactive Python.

Thank you.

**Comment by trevorstephens** [comment]
Hey there @hasakura511 ... Would you mind supplying a complete, short, simple, reproducible script either here or on a gist? Maybe use a sklearn included dataset. Best to also include your scikit-learn, numpy and scipy versions as well, as I do use those as dependencies. Try using `conda list`

Thanks for the report!

**Comment by hasakura511** [comment]
Hi Trevor,

Thank you for your reply.  GPlearn is great!  I love how you made it
fit/train like sklearn! I also appreciate the book references you put in
the code.  I am currently reading the field guide to learn how to tune the
parameters.

Please find the attached script written by Dr. Bandy, which I modified to
run the GP algo. I decided to use his script because it runs stand-alone
and it uses data similar to what I am using. I apologize the script might
be a bit long, but I wanted to recreate the environment as much as
possible.  I added the pickle lines at the end of the script. Spyder
freezes when I get to those lines. My versions are below.  I am using
virtual box, unbutu 14.14.

Also, I was wondering (if you don't mind me asking)..
1) Do you have any suggestions on how to tune the parameters to increase
accuracy on small classifications?  For example, if 70% of the y's are -1
and 30% of the y's are 1. (where y's are the predictions)  I noticed that
the algorithm usually chooses -.999 as the program, predicting all -1's.
2) With regard to using GP on other machine learning algo's, do you suggest
fitting the dataset and creating predictions from svm, logistic
regressions, etc, then appending those predictions to the original dataset
(horizontally stacking them, increasing the columns) and training the
genetic algo on the dataset with predictions?  Is this the best way?

Thank you very much.
Hidemi



Here are the versions:
hidemi@hidemi-VirtualBox:~$ conda list
# packages in environment at /home/hidemi/anaconda:
#

ipykernel                 4.1.1                    py27_0    defaults
ipython                   4.0.0                    py27_0    defaults
ipython-genutils          0.1.0                     <pip>
ipython-notebook          4.0.4                    py27_0    defaults
ipython-qtconsole         3.2.0                    py27_0    <unknown>
ipython_genutils          0.1.0                    py27_0    defaults

numpy                     1.10.1                   py27_0    defaults

scikit-image              0.11.3               np19py27_0    <unknown>
scikit-learn              0.16.1               np19py27_0    <unknown>
scipy                     0.16.0               np19py27_1    defaults

spyder                    2.3.7                    py27_1    defaults
spyder-app                2.3.5.2                  py27_0    <unknown>


On Tue, Dec 1, 2015 at 9:07 PM, Trevor Stephens <notifications@github.com>
wrote:

> Hey there @hasakura511 <https://github.com/hasakura511> ... Would you
> mind supplying a complete, short, simple, reproducible script either here
> or on a gist? Maybe use a sklearn included dataset. Best to also include
> your scikit-learn, numpy and scipy versions as well, as I do use those as
> dependencies. Try using conda list
>
> Thanks for the report!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/trevorstephens/gplearn/issues/2#issuecomment-161182925>
> .
>

# -*- coding: utf-8 -*-
"""
Figure 8.75
QTA_Chapter8_Template.py

This program is included in the book
"Quantitative Technical Analysis"
written by Dr. Howard Bandy
and published by Blue Owl Press, Inc.
Copyright 2014 Howard Bandy

Author: Howard Bandy
Blue Owl Press, Inc.
www.BlueOwlPress.com

This program is intended to be an educational tool.
It has not been reviewed.
It is not guaranteed to be error free.
Use of any kind by any person or organization
is with the understanding that the program is as is,
including any and all of its faults.
It is provided without warranty of any kind.
No support for this program will be provided.
It is not trading advice.
The programming is intended to be clear,
but not necessarily efficient.

Permission to use, copy, and share is granted.
Please include author credits with any distribution.


"""

import math
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import Quandl
import pickle
from sklearn.cross_validation import StratifiedShuffleSplit
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from gplearn.genetic import SymbolicRegressor, SymbolicTransformer

#  -------------------------------
#  Define functions.
#  These are retained in the template for reference and
#  use when needed.

def RSI(p,lb):
    # RSI technical indicator.
    # p, the series having its RSI computed.
    # lb, the lookback period, does not need to be integer.
    #     typical values in the range of 1.5 to 5.0.
    # Return is a numpy array with values in the range 0.0 to 1.0.
    nrows = p.shape[0]
    lam = 2.0 / (lb + 1.0)
    UpMove = np.zeros(nrows)
    DnMove = np.zeros(nrows)
    UpMoveSm = np.zeros(nrows)
    DnMoveSm = np.zeros(nrows)
    Numer = np.zeros(nrows)
    Denom = np.zeros(nrows)
    pChg = np.zeros(nrows)
    RSISeries = np.zeros(nrows)
    # Compute pChg in points using a loop.
    for i in range (1,nrows):
        pChg[i] = p[i] - p[i-1]    
    # Compute pChg as a percentage using a built-in method.
#    pChg = p.pct_change()
    UpMove = np.where(pChg>0,pChg,0)
    DnMove = np.where(pChg<0,-pChg,0)
    
    for i in range(1,nrows):
        UpMoveSm[i] = lam*UpMove[i] + (1.0-lam)*UpMoveSm[i-1]
        DnMoveSm[i] = lam*DnMove[i] + (1.0-lam)*DnMoveSm[i-1]
        Numer[i] = UpMoveSm[i]
        Denom[i] = UpMoveSm[i] + DnMoveSm[i]
        if Denom[i] <= 0:
            RSISeries[i] = 0.5
        else:
            RSISeries[i] =  Numer[i]/Denom[i]
    return(RSISeries)
#  -------------------------------

def ROC(p,lb):
    # Rate of change technical indicator.
    # p, the series having its ROC computed.
    # lb, the lookback period.  Typically 1.
    # Return is a numpy array with values as decimal fractions.
    # A 1% change is 0.01.
    nrows = p.shape[0]
    r = np.zeros(nrows)
    for i in range(lb, nrows):
        r[i] = (p[i]-p[i-lb])/p[i-lb]
    return(r)
#  ----------------------------------

def zScore(p,lb):
    # z score statistic.
    # p, the series having its z-score computed.
    # lb, the lookback period, an integer.
    #     the length used for the average and standard deviation.
    #     typical values 3 to 10.
    # Return is a numpy array with values as z-scores centered on 0.0.
    nrows = p.shape[0]
    st = np.zeros(nrows)
    ma = np.zeros(nrows)
    # use the pandas sliding window functions.
    st = pd.rolling_std(p,lb)
    ma = pd.rolling_mean(p,lb)
    z = np.zeros(nrows)
    for i in range(lb,nrows):
        z[i] = (p[i]-ma[i])/st[i]
    return(z)
#  ----------------------------------

def softmax(p,lb,lam):
    # softmax transformation.
    # p, the series being transformed.
    # lb, the lookback period, an integer.
    #     the length used for the average and standard deviation.
    #     typical values 20 to 252.  Be aware of ramp-up requirement.
    # lam, the length of the linear section.
    #     in standard deviations.
    #     typical value is 6.
    # Return is a numpy array with values in the range 0.0 to 1.0.
    nrows = p.shape[0]
    a = np.zeros(nrows)
    ma = np.zeros(nrows)
    sd = np.zeros(nrows)    
    sm = np.zeros(nrows)
    sq = np.zeros(nrows)
    y = np.zeros(nrows)
    for i in range(lb,nrows):
        sm[i] = sm[i]+p[i]
    ma[i] = sm[i] / lb
    for i in range(lb,nrows):
        sq[i] = (p[i]-ma[i])*(p[i]-ma[i])
    sd[i] = math.sqrt(sq[i]/(nrows-1))
    for i in range(lb,nrows):
        a[i] = (p[i]-ma[i])/((lam*sd[i])/(2.0*math.pi))
        y[i] = 1.0 / (1.0 + math.e**a[i])
    return(y)
    
#  -------------------------------    
    

def DPO(p,lb):
    # Detrended price oscillator. 
    # A high pass filter.
    # p, the series being transformed.
    # lb, the lookback period, a real number.
    # Uses pandas ewma function.
    # Return is a numpy array with values centered on 0.0.
    nrows = p.shape[0]
    ma = pd.ewma(p,span=lb)
    d = np.zeros(nrows)
    for i in range(1,nrows):
        d[i] = (p[i]-ma[i])/ma[i]
    return(d)
#  ----------------------------------
    
def numberZeros(p):
    # Counts the number of zero crossings.
    # p, the series being counted.
    # Return is an integer.  
    # The number of times p crosses up through 0.0. 
    nrows = p.shape[0]
    nz = 0
    for i in range(1,nrows):
        if (p[i-1]<0 and p[i]>=0):
            nz = nz+1
    return(nz)
    
def gainAhead(p):
    # Computes change in the next 1 bar.
    # p, the base series.
    # Return is a numpy array of changes.
    # A change of 1% is 0.01
    # The final value is unknown.  Its value is 0.0.
    nrows = p.shape[0]
    g = np.zeros(nrows)
    for i in range(0,nrows-1):
        g[i] = (p[i+1]-p[i])/p[i]
    return(g)
    
def ATR(ph,pl,pc,lb):
    # Average True Range technical indicator.
    # ph, pl, pc are the series high, low, and close.
    # lb, the lookback period.  An integer number of bars.
    # True range is computed as a fraction of the closing price.
    # Return is a numpy array of floating point values.
    # Values are non-negative, with a minimum of 0.0.
    # An ATR of 5 points on a issue closing at 50 is
    #    reported as 0.10. 
    nrows = pc.shape[0]
    th = np.zeros(nrows)
    tl = np.zeros(nrows)
    tc = np.zeros(nrows)
    tr = np.zeros(nrows)
    trAvg = np.zeros(nrows)
    
    for i in range(1,nrows):
        if ph[i] > pc[i-1]:
            th[i] = ph[i]
        else:
            th[i] = pc[i-1]
        if pl[i] < pc[i-1]:
            tl[i] = pl[i]
        else:
            tl[i] = pc[i-1]
        tr[i] = th[i] - tl[i]
    for i in range(lb,nrows):
        trAvg[i] = tr[i]            
        for j in range(1,lb-1):
            trAvg[i] = trAvg[i] + tr[i-j]
        trAvg[i] = trAvg[i] / lb
        trAvg[i] = trAvg[i] / pc[i]    
    return(trAvg)
    
#  ----------------------------------

def priceChange(p):
    nrows = p.shape[0]
    pc = np.zeros(nrows)
    for i in range(1,nrows):
        pc[i] = (p[i]-p[i-1])/p[i-1]
    return pc

#  -------------------------------------------------------    
#  -------------------------------------------------------    

#  Download data
print "\nReading data from Quandl "

ticker = "GOOG/NYSE_XLV"
dataLoadStartDate = "1998-12-22"
dataLoadEndDate = "2016-01-04"
testFirstYear = "2000"
testFinalYear = "2013"
validationFirstYear = "2014"
validationFinalYear = "2015"
iterations = 1

RSILookback = 1.0
zScoreLookback = 10
ATRLookback = 15
beLongThreshold = -0.015

model = SymbolicRegressor(population_size=5000, generations=20, stopping_criteria=0.01, comparison=True, transformer=False, p_crossover=0.7, p_subtree_mutation=0.1, p_hoist_mutation=0.05, p_point_mutation=0.1, max_samples=1, verbose=1, parsimony_coefficient=0.01, random_state=0)

qt = Quandl.get(ticker, trim_start=dataLoadStartDate,
                        trim_end=dataLoadEndDate, 
                   authtoken="aiZ4bv-njY-g1GSuYsvJ")

#print "Successfully retrieved Primary"

dataSet = qt
nrows = qt.shape[0]
print nrows, "rows retrieved.."
dataSet['Pri'] = qt.Close
dataSet['Pri_RSI'] = RSI(dataSet.Pri,RSILookback)
dataSet['Pri_ATR'] = zScore(ATR(qt.High,qt.Low,qt.Close,ATRLookback),
                          zScoreLookback)
dataSet['Pri_ATR_Y1'] = dataSet['Pri_ATR'].shift(1)
dataSet['Pri_ATR_Y2'] = dataSet['Pri_ATR'].shift(2)
dataSet['Pri_ATR_Y3'] = dataSet['Pri_ATR'].shift(3)
dataSet['priceChange'] = priceChange(dataSet['Pri'])
dataSet['priceChangeY1'] = dataSet['priceChange'].shift(1)
dataSet['priceChangeY2'] = dataSet['priceChange'].shift(2)
dataSet['priceChangeY3'] = dataSet['priceChange'].shift(3)
dataSet['Pri_RSI_Y1'] = dataSet['Pri_RSI'].shift(1)
dataSet['Pri_RSI_Y2'] = dataSet['Pri_RSI'].shift(2)
dataSet['Pri_RSI_Y3'] = dataSet['Pri_RSI'].shift(3)
dataSet['Pri_RSI_Y4'] = dataSet['Pri_RSI'].shift(4)

dataSet['gainAhead'] = gainAhead(dataSet['Pri'])
dataSet['beLong'] = np.where(dataSet.gainAhead>beLongThreshold,1,-1)

mData = dataSet.drop(['Open','High','Low','Close',
                       'Volume','Pri','gainAhead'],
                        axis=1) 

#  Select the date range to test
mmData = mData.ix[testFirstYear:testFinalYear]

datay = mmData.beLong
mmData = mmData.drop(['beLong'],axis=1)
dataX = mmData

#  Copy from pandas dataframe to numpy arrays
dy = np.zeros_like(datay)
dX = np.zeros_like(dataX)

dy = datay.values
dX = dataX.values

#  Make 'iterations' index vectors for the train-test split
sss = StratifiedShuffleSplit(dy,iterations,test_size=0.33,
                             random_state=None)

#  Initialize the confusion matrix
cm_sum_is = np.zeros((2,2))
cm_sum_oos = np.zeros((2,2))
    
#  For each entry in the set of splits, fit and predict
for train_index,test_index in sss:
    X_train, X_test = dX[train_index], dX[test_index]
    y_train, y_test = dy[train_index], dy[test_index] 

#  fit the model to the in-sample data
    model.fit(X_train, y_train)
    print '\nProgram:', model._program
    print 'R^2:    ', model.score(X_train,y_train)
    
#  test the in-sample fit    
    #y_pred_is = model.predict(X_train)
    y_pred_is = np.array(([-1 if x<0 else 1 for x in model.predict(X_train)]))
    cm_is = confusion_matrix(y_train, y_pred_is)
    cm_sum_is = cm_sum_is + cm_is

#  test the out-of-sample data
    #y_pred_oos = model.predict(X_test)
    y_pred_oos = np.array(([-1 if x<0 else 1 for x in model.predict(X_test)]))
    cm_oos = confusion_matrix(y_test, y_pred_oos)
    cm_sum_oos = cm_sum_oos + cm_oos

tpIS = cm_sum_is[1,1]
fnIS = cm_sum_is[1,0]
fpIS = cm_sum_is[0,1]
tnIS = cm_sum_is[0,0]
precisionIS = tpIS/(tpIS+fpIS)
recallIS = tpIS/(tpIS+fnIS)
accuracyIS = (tpIS+tnIS)/(tpIS+fnIS+fpIS+tnIS)
f1IS = (2.0 * precisionIS * recallIS) / (precisionIS+recallIS) 

tpOOS = cm_sum_oos[1,1]
fnOOS = cm_sum_oos[1,0]
fpOOS = cm_sum_oos[0,1]
tnOOS = cm_sum_oos[0,0]
precisionOOS = tpOOS/(tpOOS+fpOOS)
recallOOS = tpOOS/(tpOOS+fnOOS)
accuracyOOS = (tpOOS+tnOOS)/(tpOOS+fnOOS+fpOOS+tnOOS)
f1OOS = (2.0 * precisionOOS * recallOOS) / (precisionOOS+recallOOS) 

print "\n\nSymbol is ", ticker
print "Learning algorithm is ", model
print "Confusion matrix for %i randomized tests" % iterations
print "for years ", testFirstYear, " through ", testFinalYear 

print "\nIn sample"
print "     predicted"
print "      pos neg"
print "pos:  %i  %i  %.2f" % (tpIS, fnIS, recallIS)
print "neg:  %i  %i" % (fpIS, tnIS)
print "      %.2f          %.2f " % (precisionIS, accuracyIS)
print "f1:   %.2f" % f1IS

print "\nOut of sample"
print "     predicted"
print "      pos neg"
print "pos:  %i  %i  %.2f" % (tpOOS, fnOOS, recallOOS)
print "neg:  %i  %i" % (fpOOS, tnOOS)
print "      %.2f          %.2f " % (precisionOOS, accuracyOOS)
print "f1:   %.2f" % f1OOS

print "\nend of run"
print 


trained_model = pickle.dumps(model)
model = pickle.loads(trained_model)

#  /////  end  /////

**Comment by trevorstephens** [comment]
The following code works... Closing the issue as it does not appear to be related to the package. If you run this code and it fails for you, please re-open with some more details.

    from gplearn.genetic import SymbolicRegressor
    from sklearn.utils.random import check_random_state
    import numpy as np
    import pickle

    rng = check_random_state(0)

    # Training samples
    X_train = rng.uniform(-1, 1, 100).reshape(50, 2)
    y_train = X_train[:, 0]**2 - X_train[:, 1]**2 + X_train[:, 1] - 1

    # Testing samples
    X_test = rng.uniform(-1, 1, 100).reshape(50, 2)
    y_test = X_test[:, 0]**2 - X_test[:, 1]**2 + X_test[:, 1] - 1

    est_gp = SymbolicRegressor(population_size=5000,
                               generations=20, stopping_criteria=0.01,
                               comparison=False, transformer=False, 
                               p_crossover=0.7, p_subtree_mutation=0.1,
                               p_hoist_mutation=0.05, p_point_mutation=0.1,
                               max_samples=0.9, verbose=1,
                               parsimony_coefficient=0.01, random_state=0)
    est_gp.fit(X_train, y_train)

    trained_model = pickle.dumps(est_gp)
    model = pickle.loads(trained_model)
